feat: automatic type generation (#89)
tuono-labs/tuono-documentation: src/routes/documentation/integrations/typescript.mdx

import MetaTags from '@/components/MetaTags'

<MetaTags
  title="Tuono - TypeScript"
  canonical="https://tuono.dev/documentation/environment-variables"
/>

import Breadcrumbs from '@/components/Breadcrumbs'

<Breadcrumbs breadcrumbs={[{ label: 'TypeScript' }]} />

# TypeScript

## Overview

In Tuono, TypeScript is a first-class citizen.

> The `tuono` package is entirely written in TypeScript, allowing developers to leverage strong type checking to ensure program correctness when interacting with `tuono` functions and modules.

Reviewer: ```suggestion
`tuono` packages are written entirely in TypeScript, allowing developers to benefit from strong type checking that enhances code reliability and ensures correctness when using its functions and modules.
```

## Automatic Type Generation

> Typically, types need to be defined twice: first in the Rust handlers and then in the React components/routes to consume the backend payload.
> 
> Duplicating the same interfaces is error-prone and can lead to production bugs.
> 
> To simplify development, `tuono` enables automatic TypeScript type generation for all `structs` and `enums` defined in any Rust file.

Reviewer: ```suggestion
Typically, types must be defined twice: once in the Rust handlers 
and again in the React components/routes to consume the backend payload.

This duplication is error-prone and can introduce bugs.

To streamline development, `tuono` automatically generates TypeScript types for all `structs` and `enums` defined in any Rust file, 
reducing the risk of errors and ensuring consistency.
```

### The `Type` Trait

> Any struct or enum can be `derived` with the `Type` trait, and the corresponding TypeScript interface will be automatically generated on the frontend.
> 
> For example, the following struct declaration:

Reviewer: ```suggestion
Any struct or enum can be `derived` with the `Type` trait, 
which automatically generates the corresponding TypeScript interface on the frontend.

For instance, given the following struct declaration:
```

> ```rust

Reviewer: ```suggestion
```rs
```

use tuono_lib::Type;

#[derive(Type)]
struct User {
  id: String,
  email: String,
  age: u8,
  cart: Vec<Item>,
  saved: HashMap<String, Item>
}
```

> Will be available in any TypeScript file as:
> 
> ```typescript
> import type { User } from 'tuono/types'
> ```
> 
> The generated type will be mapped as:
> 
> ```typescript
> interface User {
>   id: string
>   email: string
>   age: number
>   cart: Item[]
>   saved: Record<string, Item>
> }
> ```

Reviewer: Revised and reordered the sections: first displaying the generated type, followed by the import statement.

```suggestion
The corresponding TypeScript type will be:

```ts
interface User {
  id: string
  email: string
  age: number
  cart: Item[]
  saved: Record<string, Item>
}
\`\`\`

It can be accessed via import from `tuono/types`:

```ts
import type { User } from 'tuono/types'
\`\`\`
```

Author: I like it!

### Serde Support

> To pass the payload to React, the data must be deserialized, and [serde](https://serde.rs/) is the default serializer/deserializer crate supported by `tuono`.

Reviewer: ```suggestion
To pass the payload to React, the data must be deserialized, 
and `tuono` uses [serde](https://serde.rs/) as the default serializer/deserializer.
```

> `tuono` supports all of serde's modifiers, ensuring that the generated TypeScript type matches the payload passed to the frontend.

Reviewer: ```suggestion
`tuono` supports all of serde's modifiers, ensuring the generated TypeScript type matches the payload sent to the frontend.
```

> ```rust

Reviewer: ```suggestion
```rs
```

use tuono_lib::Type;
use serde::{Serialize, Deserialize};

#[derive(Type, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct Cart {
  id: String,
  items: Vec<Item>,
  saved_items: HashMap<String, Item>,
  #[serde(rename = "total_amount")]
  cart_total_amount: String,
  #[serde(skip)]
  secret_thing: String
}
```

> Note: Manually serialized elements cannot be automatically generated.

Reviewer: Since this is a limitation it should be highlighted.

```suggestion
> ⚠️ Note: Manually serialized elements cannot be automatically generated.
```

> > This feature is strongly inspired by the [ts_rs](https://docs.rs/ts-rs/latest/ts-rs/) crate.

Reviewer: ```suggestion
### Inspiration

This feature is strongly inspired by the [ts_rs](https://docs.rs/ts-rs/latest/ts-rs/) crate.
```

Author: What about: `Credits`?

Reviewer: I chose "Inspiration" because "Credits" is typically reserved for acknowledging direct contributions or usage of someone's work. 
"Inspiration", on the other hand, more accurately reflects that this feature was influenced by the `ts_rs` crate, without directly borrowing or replicating it (the crate hasn't been included in tuono).

For example, for the following struct declaration:

```rs
use tuono_lib::Type;

#[derive(Type)]
struct User {
  id: String,
  email: String,
  age: u8,
  cart: Vec<Item>,
  saved: HashMap<String, Item>
}
```

The corresponding TypeScript type will be:

```ts
interface User {
  id: string
  email: string
  age: number
  cart: Item[]
  saved: Record<string, Item>
}
```

And it can be accessed via import from `tuono/types`:

```ts
import type { User } from 'tuono/types'
```

### Serde Support

To pass the payload to React, the data must be deserialized,
and `tuono` uses [serde](https://serde.rs/) as the default serializer/deserializer.

`tuono` supports all of serde's modifiers, ensuring that the generated TypeScript
type matches the payload sent to the frontend.

```rs
use tuono_lib::Type;
use serde::{Serialize, Deserialize};

#[derive(Type, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct Cart {
  id: String,
  items: Vec<Item>,
  saved_items: HashMap<String, Item>,
  #[serde(rename = "total_amount")]
  cart_total_amount: String,
  #[serde(skip)]
  secret_thing: String
}
```

> ⚠️ Note: Manually serialized elements cannot be automatically generated.

### Credits

> This feature is strongly inspired by the [ts_rs](https://docs.rs/ts-rs/latest/ts-rs/) crate.

Author: Updated
